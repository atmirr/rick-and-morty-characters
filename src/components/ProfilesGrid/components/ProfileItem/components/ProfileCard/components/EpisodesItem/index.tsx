import React, { useState } from 'react';
import styled from 'styled-components';
import EpisodesModal from './components/EpisodesModal';

const Wrapper = styled.div`
  display: flex;
  padding: 6px 1px;
`;

const Label = styled.span`
  font-size: 16px;
  font-weight: 500;
  color: #7c7c7c;
`;

const Value = styled.span`
  padding-left: 3px;
  color: #222222;
`;

const MoreLabel = styled.a`
  font-weight: 500;
  color: #1976d2;
  cursor: pointer;
`;

type PropTypes = {
  label?: string;
  children: string[];
};

function EpisodesItem({
  label = 'Episodes',
  children: episodes,
}: PropTypes): React.ReactElement {
  const firstEpisode = episodes?.[0];
  const episodesLength = episodes?.length;
  const moreEpisodesText = `and ${episodesLength - 1} more`;
  const [isModalOpen, setIsModalOpen] = useState(false);
  const handleOpenModal = () => setIsModalOpen(true);
  const handleCloseModal = () => setIsModalOpen(false);
  return (
    <>
      <Wrapper data-testid="episodes-item">
        <Label>{label}:</Label>
        <div>
          <Value>{firstEpisode}</Value>{' '}
          {episodesLength > 1 && (
            <MoreLabel onClick={handleOpenModal}>{moreEpisodesText}</MoreLabel>
          )}
        </div>
      </Wrapper>
      {isModalOpen && (
        <EpisodesModal
          episodes={episodes}
          handleCloseModal={handleCloseModal}
        />
      )}
    </>
  );
}

export default EpisodesItem;
