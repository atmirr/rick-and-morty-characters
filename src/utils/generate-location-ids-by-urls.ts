export const generateLocationIdsByUrls = (urls: string[]): string[] =>
  urls.reduce((result: string[], url: string) => {
    const splittedUrl = url.split('/');
    const id = splittedUrl[splittedUrl.length - 1];
    return id ? [...result, id] : [...result];
  }, []);
