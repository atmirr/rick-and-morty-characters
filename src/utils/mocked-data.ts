export const characters = [
  {
    id: 1,
    name: 'Rick Sanchez',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:48:46.250Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/1',
  },
  {
    id: 2,
    name: 'Morty Smith',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:50:21.651Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
      'https://rickandmortyapi.com/api/episode/3',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/2',
  },
];

export const episodes = [
  'Rixty Minutes',
  'Something Ricked This Way Comes',
  'Close Rick-counters of the Rick Kind',
  'Ricksy Business',
  'A Rickle in Time',
];

export const locations = [
  {
    name: 'Citadel of Ricks',
    type: 'Space station',
    dimension: 'unknown',
  },
  {
    name: 'Earth (Replacement Dimension)',
    type: 'Planet',
    dimension: 'Replacement Dimension',
  },
];
