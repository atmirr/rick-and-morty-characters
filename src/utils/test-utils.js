import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockedInitialState = {
  characters: { items: [], nextPage: null, error: null, isLoading: false },
  episodes: { items: [], nextPage: null, error: null, isLoading: false },
  locations: { items: [], error: null, isLoading: false },
};

function render(
  ui,
  {
    initialState = mockedInitialState,
    store = configureStore()(initialState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
