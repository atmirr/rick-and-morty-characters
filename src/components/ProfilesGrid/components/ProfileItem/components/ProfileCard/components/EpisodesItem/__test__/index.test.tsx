import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { screen, render, fireEvent } from 'src/utils/test-utils';
import { episodes } from 'src/utils/mocked-data';
import EpisodesItem from '../index';

const props = {
  label: 'Chapters',
  children: episodes,
};

describe('<EpisodesItem />', () => {
  it('should renders the component', () => {
    const { container } = render(<EpisodesItem />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should support default value for the label', () => {
    render(<EpisodesItem>{props.children}</EpisodesItem>);
    expect(screen.queryAllByText(/Episodes:/i).length).toBe(1);
  });

  it('should support custom value for the label', () => {
    render(<EpisodesItem label={props.label}>{props.children}</EpisodesItem>);
    expect(screen.queryAllByText(`${props.label}:`).length).toBe(1);
  });

  it('should not display "more episodes" label when the children array has just one key', () => {
    render(<EpisodesItem>{[props.children[0]]}</EpisodesItem>);
    expect(screen.queryAllByText(/(and )+([0-9]*)+( more)/i).length).toBe(0);
    expect(screen.queryAllByText(props.children[0]).length).toBe(1);
  });

  it('should display "more episodes" label when the children array has more than one key', () => {
    render(<EpisodesItem>{props.children}</EpisodesItem>);
    expect(screen.queryAllByText(props.children[0]).length).toBe(1);
    expect(screen.queryAllByText(/and 4 more/i).length).toBe(1);
  });

  it('should open the episodes modal when user clicks on the "and %count% more" link', () => {
    render(<EpisodesItem>{props.children}</EpisodesItem>);
    const moreLink = screen.getByText(/and 4 more/i);
    fireEvent.click(moreLink);
    expect(screen.getAllByTestId('modal').length).toBe(1);
  });
});
