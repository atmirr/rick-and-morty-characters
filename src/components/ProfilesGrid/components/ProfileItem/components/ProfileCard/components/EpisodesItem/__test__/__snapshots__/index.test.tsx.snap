// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EpisodesItem /> should renders the component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 6px 1px;
}

.c1 {
  font-size: 16px;
  font-weight: 500;
  color: #7c7c7c;
}

.c2 {
  padding-left: 3px;
  color: #222222;
}

<div
  class="c0"
  data-testid="episodes-item"
>
  <span
    class="c1"
  >
    Episodes
    :
  </span>
  <div>
    <span
      class="c2"
    />
     
  </div>
</div>
`;
